File: Week2/ArrayProblem/ThreeSum.java
import java.util.List;
import java.util.HashSet;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Set;
import java.util.Arrays;

class ThreeSum {
/*
 * Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, 
 * and nums[i] + nums[j] + nums[k] == 0.  Solution set must not contain duplicate triplets.
 * https://leetcode.com/problems/3sum/
 * T(c) -> O(N^2 * log(no. of unique triplets)) // time required for storing inside set datastructure.
 * S(c) -> O(2 * no. of the unique triplets) + O(N)  // first is for list and set, 2nd is for set.
 */
    static List<List<Integer>> threeSum(int[] nums) {
        /* for storing unique arrays */
        Set<List<Integer>> hashset = new HashSet<>();
        
        for(int i=0;i<nums.length;i++)
        {
            Set<Integer> tempSet = new HashSet<>();
            for(int j=i+1;j<nums.length;j++)
            {
                /* calculating third element, i+j+k = 0 => k = -i-j => k = -(i+j) */
                int third = -(nums[i]+nums[j]);
                
                /* if hashset contains that element */
                if(tempSet.contains(third))
                {
                    List<Integer> tempList = Arrays.asList(nums[i],nums[j],third);
                    Collections.sort(tempList);
                    hashset.add(tempList);
                }
                /* add number in hashset */
                tempSet.add(nums[j]);
            }
        }
        
        List<List<Integer>> result = new ArrayList<>(hashset);
        return result;
    }


    public static void main(String[] args) {
        int arr[] = {-1,0,1,2,-1,-4};
        List<List<Integer>> result = threeSum(arr);
        for(List<Integer> res:result)
        {
            for(int elem: res)
            {
                System.out.print(elem+" ");
            }
            System.out.println();
        }
    }
}

File: Week2/Python/ArrayProblem/ThreeSum.py
"""
    Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, 
    and nums[i] + nums[j] + nums[k] == 0.
    Notice that the solution set must not contain duplicate triplets.

    https://leetcode.com/problems/3sum/
"""


def threeSum(nums):
    """Time Limit issue"""
    """ T(c) -> O(n^3) , S(c) -> O(n) """
    result = []
    
    for i in range(len(nums)):
        for j in range(i+1, len(nums)):
            for k in range(j+1, len(nums)):
                
                if nums[i] + nums[j] + nums[k] == 0:
                    
                    # create triplet and sort it to discard duplicates
                    triplet = [nums[i], nums[j], nums[k]]
                    triplet.sort()
                    
                    if triplet not in result:
                        result.append(triplet)
    
    return result


def threeSumBetter(nums):
    """ T(c) -> O(n^2) , S(c) -> O(n) """
    result = set()
    
    for i in range(len(nums)):
        seen = set()
        for j in range(i+1, len(nums)):
            another_num = -nums[i] - nums[j]
            
            if another_num in seen:
                # create triplet and sort it to discard duplicates
                triplet = tuple(sorted((nums[i], nums[j], another_num)))
                result.add(triplet)
            
            # add current element in seen set
            seen.add(nums[j])


    return [list(triple) for triple in result]


def threeSumOptimal(nums):
    """ T(c) -> O(n^2) , S(c) -> O(n) """

    nums.sort()
    
    result = []
    n = len(nums)
    
    # fix i pointer
    for i in range(n):

        # skip duplicates of i
        if i > 0 and nums[i] == nums[i-1]:
            continue
        
        # take 2 pointers left and right
        left = i + 1
        right = n -1

        while left < right:
            total = nums[i] + nums[left] + nums[right]
            
            if total == 0:
                result.append([nums[i], nums[left], nums[right]])
                
                # move both pointers and avoid duplicates
                while left < right and nums[left] == nums[left+1]:
                    left += 1
                
                while left < right and nums[right] == nums[right-1]:
                    right -= 1
                    
                left += 1
                right -= 1
                
            # move left pointer to increase sum
            elif total < 0:
                left += 1

            # move right pointer to decrease sum
            else:
                right -= 1

    return result


arr = [-1,0,1,2,-1,-4]
print(threeSum(arr))

print(threeSumBetter(arr))

print(threeSumOptimal(arr))
